### PR TITLE
grass.script: Add MapsetSession for runs in other mapsets

### PR DESCRIPTION
**Pull request by wenzeslaus**
The new MapsetSession object uses the current db/location (project) and changes mapset (subproject) in a new separate environment. A new mapset is created if requested. MapsetSession can be used as a context manager. The original use case is bulk imports.

This combines existing create_mapset and create_environment and adds the require/create/ensure logic, one interface, and resource handling. The underlying functionality is general enough and available as a new require_create_ensure_mapset function.

The MapsetSession interface is the same as the session handle returned from grass.script.setup.init which now gets a new env property to keep the interface unified and to allow for session-object-agnostic code in the tests.

A separate TemporaryMapsetSession adds a temporary mapset handling which shares most of the API, but the implementation is very different since the mapset never exists and is always created. There the context manager creates and deletes the temporary mapset within the with-statement. The expected usage is one-off computations which are part of a larger process in GRASS GIS or outside of it.

Other possible names are MapsetSubSession and SubMapsetSession.

### CODE REVIEW

**Comment by wenzeslaus** [comment]
Now this allows also temporary mapsets:

```python
with gs.setup.init("~/grassdata/nc_spm_08_grass7"):
    with TemporaryMapsetSession() as session:
        # Do some work, import, export, get statistics, ...
        gs.run_command("g.region", res=10, rows=100, cols=200, env=session.env)
        gs.run_command(
            "r.surf.random", output="uniform_random", min=1, max=10, env=session.env
        )
        print(
            gs.parse_command(
                "r.univar", map="uniform_random", flags="g", env=session.env
            )["max"]
        )
    gs.run_command("g.mapsets", flags="l")
# Now you have whatever was in nc_spm_08_grass7, but nothing else.
```

**Comment by wenzeslaus** [comment]
This is now ready to be merged. The description is updated accordingly.

It now includes temporary mapset session which handles the case where mapset is created at the beginning and deleted at the end of the processing.

I'm adding it under new subpackage _grass.experimental_. I hope that this package will allow to merge code which is fully functional, but does not have all corner cases or API resolved. We did something similar already with _grass.jupyter_ which was included in v8.0, but marked experimental and it was promoted and included in documentation only in 8.2 after additional changes. In general, this should allow use and testing by advanced users without creating high expectations or committing to a specific API or behavior.

File: python/grass/experimental/mapset.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    variables for the session to subprocesses:
+
+    >>> with MapsetSession(name, ensure=True) as session:
+    ...     run_command("r.surf.fractal", output="surface", env=session.env)
+
+    This session expects an existing GRASS runtime environment.
+
+    The name argument is positional-only.
+
+    .. versionadded:: 8.3
```
Comment (landam) [inline python/grass/experimental/mapset.py:39]:
```suggestion
    .. versionadded:: 8.4
```

File: python/grass/experimental/mapset.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    Standard use of the object is to use it as a context manager, i.e., create it
+    using the ``with`` statement. Then use its *env* property to pass the environment
+    variables for the session to subprocesses:
+
+    >>> with MapsetSession(name, ensure=True) as session:
+    ...     run_command("r.surf.fractal", output="surface", env=session.env)
+
+    The name argument is positional-only.
+
+    .. versionadded:: 8.3
```
Comment (landam) [inline python/grass/experimental/mapset.py:143]:
```suggestion
    .. versionadded:: 8.4
```

**Review by landam — APPROVED** [review]

**Comment by echoix** [comment]
How smooth does the transition from experimental to non experimental would be?

**Comment by wenzeslaus** [comment]
The transition from experimental to non-experimental in code means the following:

1. Figuring out whether the API really makes sense. This might be best done with more general session in mind which connects on project/location-level rather than mapset. I have some work around that lined up for the following months, but most likely after 8.4.0.
2. Moving the code to its new home (e.g., grass.script.setup or new grass.script.session)
3. Optional: Providing the new code through the experimental package just to be nice to whoever started to use that. (And removing that one release after.)

**Comment by echoix** [comment]
> 3. Optional: Providing the new code through the experimental package just to be nice to whoever started to use that. (And removing that one release after.)

Is that last part documented somewhere? I think that's the important part, just like when we officialised what python versions we support and when can start moving on from old patterns that are required for backwards compatibility

**Comment by neteler** [comment]
> what python versions we support

Please see:

RFC 8: Python Language Support
https://github.com/OSGeo/grass/blob/main/doc/development/rfc/python_language_support.md

**Comment by echoix** [comment]
Yes, that's the document I'm talking about. Is there something similar to how we handle removing experimental that is graduated to general availability?

**Comment by wenzeslaus** [comment]
> Is there something similar to how we handle removing experimental that is graduated to general availability?

_grass.experimental_ is new and we don't have a separate document for it except what is in this PR: "Experimental code, all can change" Do you think more is needed somewhere?

**Pull request merged**